crates/core/src/host/wasmtime/wasm_instance_env.rs

```diff
@@ -1055,15 +1131,59 @@ impl WasmInstanceEnv {
 
             // Destroy the source if exhausted, or advance `cursor`.
             if remainder.is_empty() {
-                env.call_reducer_args = None;
+                env.free_bytes_source(source);
                 Ok(-1i32)
             } else {
-                *cursor += can_read_len;
+                bytes_source.cursor += can_read_len;
                 Ok(0)
             }
         })
     }
 
+    /// Read the remaining length of a [`BytesSource`] and write it to `out`.
+    ///
+    /// Note that the host automatically frees byte sources which are exhausted.
+    /// Such sources are invalid, and this method will return an error when passed one.
+    /// Callers of [`Self::bytes_source_read`] should check for a return of -1
+    /// before invoking this function on the same `source`.
+    ///
+    /// Also note that the special [`BytesSourceId::INVALID`] (zero) is always invalid.
+    /// Callers should check for that value before invoking this function.
+    ///
+    /// # Traps
+    ///
+    /// Traps if:
+    ///
+    /// - `out` is NULL or `ou` is not in bounds of WASM memory.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error:
+    ///
+    /// - `NO_SUCH_BYTES`, when `source` is not a valid bytes source.
+    ///
+    /// If this function returns an error, `out` is not written.
+    pub fn bytes_source_remaining_length(caller: Caller<'_, Self>, source: u32, out: WasmPtr<u32>) -> RtResult<i32> {
+        Self::cvt_custom(caller, AbiCall::BytesSourceRemainingLength, |caller| {
+            let (mem, env) = Self::mem_env(caller);
+
+            let Some(bytes_source) = env.bytes_sources.get(&BytesSourceId(source)) else {
+                return Ok(errno::NO_SUCH_BYTES.get().into());
+            };
+
+            let total_len = bytes_source.bytes.len();
```

Comment:
If we just let the `bytes::Bytes` instance track the offset, this can be simplified to:

```
let remaining: u32 = bytes_source.bytes.len().try_into()
    .context("Bytes object in `BytesSource` had length greater than range of u32")?;
```